OAuth 2.0 BETA branch

larsjohansen-okta opened the pull request:
This is the beta-doc PR for OAuth 2.0. We will use it to incorporate all doc changes until we beta private resource server. The list below keeps track of the sections added/modified compared to the regular doc.

- Added OAuth 2.0 beta page
- Added custom scopes and claims to the beta page
- Added service client and client credentials flow

Reviewers: @federations-okta 

Note: We don't intend to merge this until we actually beta OAuth Private Resource Server.

cwu-okta commented on _source/docs/api/resources/oauth2-beta.md:
```diff
@@ -0,0 +1,441 @@
+If the request that generates the access token contains any custom scopes, those scopes will be part of the <b>scp</b> claim together with the scopes provided from the [OIDC specification](http://openid.net/specs/openid-connect-core-1_0.html).
+
+#### Custom claims
+All configured custom claims will be part of the access token and evaluated for the user.
```
Should you indicate where/how you can configure custom scopes and claims?

larsjohansen-okta replied:
Added it to the "Custom scopes and claims" section.

cwu-okta commented on _source/docs/api/resources/oauth2-beta.md:
```diff
@@ -0,0 +1,441 @@
+A client may only revoke a token generated for that client.
+
+The client can authenticate by providing <em>client_id</em> and <em>client_secret</em> as a part of the URL-encoded form parameters (as described in table above),
+or it can use basic authentication by providing the <em>client_id</em> and <em>client_secret</em> as an Authroization header using the Basic auth scheme. 
```
Authroization -> Authorization

larsjohansen-okta replied:
Fixed.

cwu-okta commented on _source/docs/api/resources/oauth2-beta.md:
```diff
@@ -0,0 +1,441 @@
+jti         | The identifier of the token.                                                                        | String     |
+device_id   | The ID of the device associated with the token                                                      | String     |
+
+Custom scopes are added to the list of scopes. Note that this feature is in Beta status.
```
Should there be an example of the custom scopes returned in a request?

royalchan-okta replied:
They are just string values, maybe add an extra Response Example for custom scopes so that you can compare

royalchan-okta commented:
Not related to your change but since you are fixing some `oauth2.md` bugs, the <em> on line 132 and 140 in the response example should be removed

<img width="373" alt="screen shot 2016-06-08 at 10 22 11 pm" src="https://cloud.githubusercontent.com/assets/13244399/15919572/7e825766-2dc7-11e6-8cab-3f3067bd197b.png">

royalchan-okta commented on _source/docs/api/resources/oauth2-beta.md:
```diff
@@ -0,0 +1,441 @@
+##### Token Authentication Methods
+
+The client can authenticate by providing <em>client_id</em> and <em>client_secret</em> as a part of the URL-encoded form parameters (as described in table above),
+or it can use basic authentication by providing the <em>client_id</em> and <em>client_secret</em> as an Authroization header using the Basic auth scheme. 
```
Authroization

mystiberry-okta replied:
I'll find it and fix in oauth2.md

larsjohansen-okta replied:
Fixed it here as well.

royalchan-okta commented:
since this is a beta branch, you should just add oauth2-beta.md to the docs.yml so you can preview the html

mystiberry-okta commented:
I'm confused. We already have http://developer.okta.com/docs/api/resources/oauth2.html published as EA. Stephen Lee was asking about it, too. Are we repeating content in the two docs to make it easy for the beta customers? Just want to make sure everything's okay...

royalchan-okta commented:
@mystiberry-okta This is the actual beta branch that we originally discussed for the purposes of doc'ing all the OAuth changes that aren't EA yet.

cwu-okta commented:
@mystiberry-okta this is a branch which will make available when oauth private resource server goes to beta (targeted for end of June)

mystiberry-okta commented:
So we'll still have one OAuth 2.0 file when we're done, yes?

cwu-okta commented:
When we are in oauth private resource server beta there will be 2 OAuth2.0 files. Once we go into OAuth private resource server EA there will be a single OAuth 2.0 file.

mystiberry-okta commented:
Perfect. We just have to be careful about catching any changes in the older file for the duplicated content in the newer file. Thanks!

larsjohansen-okta commented:
@royalchan-okta we cannot add it to the ```docs.yml``` since when we merge this (when private resource server goes beta) we don't want the beta page to show up on the page other than if you are provided the link directly.

larsjohansen-okta commented:
@royalchan-okta could you explain what should be removed in the example you took a screenshot of?

larsjohansen-okta commented:
![image](https://cloud.githubusercontent.com/assets/9053598/15952519/6f8706f8-2e76-11e6-9c97-19ef90f8cb1b.png)

royalchan-okta commented:
@larsjohansen-okta the \<em\> should be removed, it actually turned it into markup in github and italicized the words...

larsjohansen-okta commented:
@royalchan-okta Gotcha. I fixed that in the beta-doc, but forgot to fix it in oauth2.md. It's now fixed in both. It was also missing a ```&``` between token_type and state parameter.

royalchan-okta commented on _source/docs/api/resources/oauth2-beta.md:
```diff
@@ -0,0 +1,462 @@
+#### Response Example (Success)
+
+~~~http
+HTTP/1.1 200 OK
```
I forgot to fix this, but it turns out this returns a `204 No Content`

larsjohansen-okta replied:
Fixed.

royalchan-okta commented:
LGTM, just that one minor fix left

mystiberry-okta commented:
Since we can't push this until ready for Beta, the oauth2.md fixes won't get pushed either. Unless there's some git ninja tricks for only pushing one file?

larsjohansen-okta commented:
@mystiberry-okta you're right, and no there's no way we can push this single file. We can create a new branch and copy the changes for this single file. I suggest we do that so we get the bug fixes in as soon as possible, so I created a PR with the changes for oauth2.md only here https://github.com/okta/okta.github.io/pull/270. Please review so we can merge those typos and minor bug fixes.

larsjohansen-okta commented on _source/docs/api/resources/oauth-clients.md:
```diff
@@ -63,10 +63,10 @@ Client applications have the following properties:
 | client_secret              | OAuth 2.0 client secret string (used for confidential clients)    | String                                                                 | TRUE     | TRUE   | TRUE     |
 | logo_uri                   | URL string that references a logo for the client                  | String                                                                 | TRUE     | FALSE  | FALSE    | 
-| application_type           | The type of client application                                    | `web`, `native`, or `browser`                                          | TRUE     | TRUE   | TRUE     |
+| application_type           | The type of client application                                    | `web`, `native`, `browser` or `service`                                | TRUE     | TRUE   | TRUE     |
```
Do we need to doc that the service app type is in beta once we merge this PR?

yuliu-okta replied:
Added both here and in the token endpoint.

mystiberry-okta replied:
Doc looks fantastic!

larsjohansen-okta commented on _source/docs/api/resources/oauth2-beta.md:
```diff
@@ -262,7 +265,7 @@ The admin can configure custom scopes and claims via the UI <a href="...link to
 
 #### Custom claims
-All configured custom claims will be part of the access token and evaluated for the user.
+All configured custom claims will be part of the access token. And they will be evaluated for the user if the grant type is not <em>client_credentials</em>.
```
I would rephrase it like this: ```All configured custom claims will be part of the access token. They will be evaluated for the user if the grant type is not <em>client_credentials</em>.```

yuliu-okta replied:
fixed

royalchan-okta commented:
Superseded by https://github.com/okta/okta.github.io/pull/299
